### ISSUE TITLE
Inconsistent use of `nelec` property for pbc calculations.

### ISSUE THREAD

**Issue by verena-neufeld** [pyscf/pyscf#2035]
The `nelec` property as defined in https://github.com/pyscf/pyscf/blob/009f1d3899fbdd809fb4d73c9b06a7ab7b4b5e56/pyscf/pbc/gto/cell.py#L952 assumes that `self.nelectron`, `self.spin`, `nalpha` and `nbeta` are all per cell I think. For k point SCF calculations, this function is redefined to assume that number of electrons, spin, nalpha, nbeta scale with number of k points, but this redefinition happens in the SCF files, see e.g. https://github.com/pyscf/pyscf/blob/009f1d3899fbdd809fb4d73c9b06a7ab7b4b5e56/pyscf/pbc/scf/kuhf.py#L394 . Here, I think, `cell.spin` scales with number of k points. This feels easy to get confused about (one such function in cell.py, others in the SCF .py files, assuming different scaling with k points). 

In this TDSCF test https://github.com/pyscf/pyscf/blob/009f1d3899fbdd809fb4d73c9b06a7ab7b4b5e56/pyscf/pbc/tdscf/test/test_kuhf.py#L75 , this does actually lead to a potential bug. Here, `cell.spin` (as the comment says) scales with k points but the `nelec` function in `cell.py` is being called, which assumes spin per k point I think. When `cell.spin` is set to `12` for example here (line 88) and correspondingly `kpts = cell.make_kpts((6,1,1))` in line 92, the `nbeta` given by `nelec` is now negative (-1) which is unphysical.

Please let me know if I misunderstood the units. In either case, some clarifications are probably helpful here as it is easy to introduce bugs.

**Comment by tberkel** [comment]
If you could make a minimal working example that shows a clear bug, that would make this easier to confirm and fix.

**Comment by verena-neufeld** [comment]
Of course, thanks! I've modified the previously mentioned test (pyscf/pyscf/pbc/tdscf/test/test_kuhf.py):

```
import numpy as np
from pyscf.pbc import gto, scf

cell = gto.Cell()
cell.atom = '''
O          0.00000        0.00000        0.11779
H          0.00000        0.75545       -0.47116
H          0.00000       -0.75545       -0.47116
'''
cell.spin = 12   # TOTAL spin in the corresponding supercell
cell.a = np.eye(3) * 15
cell.basis = 'sto-3g'
cell.build()
print("nalpha, nbeta using cell.nelec: ", cell.nelec)
kpts = cell.make_kpts((6,1,1))
mf = scf.KUHF(cell, kpts=kpts).rs_density_fit(auxbasis='weigend')
mf.kernel()
print("nalpha, nbeta using mf.nelec: ", mf.nelec)
```

Giving:
`nalpha, nbeta using cell.nelec:  (11, -1)` and `nalpha, nbeta using mf.nelec:  (36, 24)` which is not consistent and gives a negative `nbeta` for the `cell.nelec` version.

**Comment by fishjojo** [comment]
I think it's more consistent to define spin and nelec only for unit cell.

**Comment by tberkel** [comment]
Thanks @verena-neufeld for the MWE. To @fishjojo, I agree that would be more consistent, but it would prevent studying a large supercell (via k-point sampling) with only a few unpaired electrons.

**Comment by verena-neufeld** [comment]
We could allow fractional `cell.spin` entries and fractional `nalpha` and `nbeta`. Then the user could specify `cell.spin` per unit cell, i.e. per k point. That might work?

**Comment by tberkel** [comment]
The same issue exists with `nelec`. How do I study a 2-cell supercell (via 2 k-points) with 3 electrons? The number of electrons per cell would have to be fractional (1.5). In previous applications, we've gotten around all of these issues by just building the actual supercell, but perhaps there's a better alternative. We could see what other periodic codes do.

**Comment by verena-neufeld** [comment]
It might be ok to have a fractional number of electrons (also relevant for calculations with smearing) and when inputed into functions that require integer number of electrons, it could be checked whether nkpts*nelectrons is an integer.

**Comment by verena-neufeld** [comment]
Thanks Zhi-Hao! My suggestion would be to define spin and number of electrons per cell, allowing fractions, but deprecating the cell.spin keyword for a while, and instead using maybe cell.spin_pkpt or something to make sure other codes don't give wrong answers. I think the current behaviour can lead to bugs and isn't intuitive (I think at least) so I'm personally in favour of a consistent definition (I propose per cell). But of course it would lead to fractional number of electrons and spin.

**Comment by gkc1000** [comment]
My memory of this convention is that it was set up in the very beginning when we were trying to make the periodic code behave like the molecular code, and for that reason we needed to guarantee nelec to be an integer. (This was even before smearing was implemented). Since after smearing, nelec can be arbitrary (in either the molecule or solid), it seems ok to allow cell.nelec etc. to be fractional. This will certainly break quite a few input files.
